src/tree/mod.rs

```diff
@@ -1,103 +1,430 @@
-                    SYNC_TO_HEAD_SLEEP_SECONDS,
-                ))
-                .await;
+                    let mut identity_tree = identity_tree.write().await;
+                    // We can use expect here because the root will always be in tree updates before the root is bridged to other chains
+                    let root_nonce = identity_tree
+                        .roots
+                        .get(&bridged_root)
+                        .expect("Could not get root update");
+                    let new_root = Root {
+                        hash: bridged_root,
+                        nonce: *root_nonce,
+                    };
+
+                    // If the update is for the chain with the oldest root, apply the updates to the tree
+                    if chain_id == *oldest_root.0 {
+                        identity_tree.apply_updates_to_root(oldest_root.1);
+                    }
+
+                    // Update chain state with the new root
+                    chain_state.insert(chain_id, new_root);
+                }
             }
         })
     }
+
+    async fn get_latest_roots(&self) -> eyre::Result<HashMap<u64, Hash>> {
+        let mut tree_data = vec![];
+
+        for bridged_tree in self.bridged_tree_manager.iter() {
+            let bridged_world_id = IBridgedWorldID::new(
+                bridged_tree.address,
+                bridged_tree.block_scanner.middleware.clone(),
+            );
+
+            tree_data.push((bridged_tree.chain_id, bridged_world_id));
+        }
+
+        tree_data.push((
+            self.canonical_tree_manager.chain_id,
+            IBridgedWorldID::new(
+                self.canonical_tree_manager.address,
+                self.canonical_tree_manager.block_scanner.middleware.clone(),
+            ),
+        ));
+
+        let futures = tree_data.iter().map(|(chain_id, contract)| async move {
+            let root: U256 = contract.latest_root().await?;
+
+            eyre::Result::<_, eyre::Report>::Ok((
+                *chain_id,
+                Uint::<256, 4>::from_limbs(root.0),
+            ))
+        });
+
+        let roots = futures::future::try_join_all(futures)
+            .await?
+            .into_iter()
+            .collect::<HashMap<u64, Hash>>();
+
+        Ok(roots)
+    }
+
+    #[instrument(skip(self))]
+    pub async fn sync_to_head(&self) -> eyre::Result<()> {
+        // Update the last synced block for each bridged tree to the current block
+        for bridged_tree in self.bridged_tree_manager.iter() {
+            let current_block = bridged_tree
+                .block_scanner
+                .middleware
+                .get_block_number()
+                .await?
+                .as_u64();
+
+            bridged_tree
+                .block_scanner
+                .last_synced_block
+                .store(current_block, Ordering::SeqCst);
+        }
+
+        // Get all logs from the canonical tree from the last synced block to the chain tip
```

Comment:
This actually gets all logs from that last synced to `min(current block, last synced + window size)`. So this function won't actually sync to head